src/parser/statement.js

```
@@ -303,35 +303,34 @@ pp.parseForStatement = function (node) {
     return this.parseFor(node, init);
   }
 
+  if (this.hasPlugin("lightscript") && (!forAwait)) {
+    // idx, elem, key, or val begins a LS-enhanced loop.
+    if (
+      (this.isContextual("idx") || this.isContextual("elem") || this.isContextual("key") || this.isContextual("val"))
+    ) {
+      return this.parseEnhancedForIn(node);
+    }
+  }
+
   const refShorthandDefaultPos = { start: 0 };
   const init = this.parseExpression(true, refShorthandDefaultPos);
   if (this.match(tt._in) || this.isContextual("of")) {
+    if (this.hasPlugin("lightscript") && (!init.isNowAssign)) {
+      this.raise(this.state.lastTokStart, "For-in requires `const` or `now`.");
```

whoops, my earlier comment about the error message may also have been swallowed... let me dig that up...

---

Might be worth suggesting the new options too...

> for-in requires a variable keyword: idx, elem, key, or val to used LightScript enhanced loops, now to reassign a preexisting variable, or const, let, or var to declare a new one with a traditional for-in loop.

Meanwhile, for-of needs the simpler error message you have now (though let and var are allowed too)

---

```
      if (this.match(tt._in)) {
        this.raise(this.state.lastTokStart, "for-in requires a variable qualifier: `now` to reassign an existing variable, or `const`, `let`, `var` to declare a new one. Use `idx` or `elem` to iterate an array. Use `key` or `val` to iterate an object.");
      } else {
        this.raise(this.state.lastTokStart, "for-of requires a variable qualifier: `now`, `const`, `let` or `var`.");
      }
```
?

---

Perfect!!